// @tarojs/plugin-framework-vue3/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var helper = require('@tarojs/helper');
var shared = require('@tarojs/shared');
var template = require('@tarojs/shared/dist/template');
var runnerUtils = require('@tarojs/runner-utils');
var lodash = require('lodash');

function getLoaderMeta() {
    return {
        importFrameworkStatement: `
import { h, createApp } from 'vue'
`,
        mockAppStatement: `
const App = createApp({})
`,
        frameworkArgs: 'h, config',
        creator: 'createVue3App',
        creatorLocation: '@tarojs/plugin-framework-vue3/dist/runtime',
        importFrameworkName: 'h',
        isNeedRawLoader: true,
        extraImportForWeb: '',
        execBeforeCreateWebApp: '',
    };
}

function getVueLoaderPath() {
    try {
        return require.resolve('vue-loader', {
            paths: [process.cwd()]
        });
    }
    catch (error) {
        // eslint-disable-next-line no-console
        console.log(helper.chalk.yellow('找不到 vue-loader，请先安装。'));
        process.exit(1);
    }
}

function modifyH5WebpackChain(ctx, chain, config) {
    var _a;
    // vue3 tsx 使用原生组件
    setVueLoader$1(chain, config);
    setLoader$1(chain);
    setTaroApiLoader(chain);
    const { isBuildNativeComp = false } = ((_a = ctx.runOpts) === null || _a === void 0 ? void 0 : _a.options) || {};
    const externals = {};
    if (isBuildNativeComp) {
        // Note: 该模式不支持 prebundle 优化，不必再处理
        externals.vue = 'vue';
    }
    chain.merge({ externals });
}
function setVueLoader$1(chain, config) {
    var _a, _b;
    const vueLoaderPath = getVueLoaderPath();
    // plugin
    const { VueLoaderPlugin } = require(vueLoaderPath);
    chain
        .plugin('vueLoaderPlugin')
        .use(VueLoaderPlugin);
    const compilerOptions = ((_a = config.vueLoaderOption) === null || _a === void 0 ? void 0 : _a.compilerOptions) || {};
    // loader
    const vueLoaderOption = Object.assign(Object.assign({ transformAssetUrls: {
            video: ['src', 'poster'],
            'live-player': 'src',
            audio: 'src',
            source: 'src',
            image: 'src',
            'cover-image': 'src',
            'taro-video': ['src', 'poster'],
            'taro-live-player': 'src',
            'taro-audio': 'src',
            'taro-source': 'src',
            'taro-image': 'src',
            'taro-cover-image': 'src'
        } }, ((_b = config.vueLoaderOption) !== null && _b !== void 0 ? _b : {})), { compilerOptions: Object.assign(Object.assign({}, compilerOptions), { 
            // https://github.com/vuejs/vue-next/blob/master/packages/compiler-core/src/options.ts
            nodeTransforms: [(node) => {
                    if (node.type === 1 /* ELEMENT */) {
                        node = node;
                        const nodeName = node.tag;
                        if (runnerUtils.DEFAULT_Components.has(nodeName)) {
                            node.tag = `taro-${nodeName}`;
                            node.tagType = 1; /* 0: ELEMENT, 1: COMPONENT */
                        }
                    }
                }] }) });
    chain.module
        .rule('vue')
        .test(helper.REG_VUE)
        .use('vueLoader')
        .loader(vueLoaderPath)
        .options(vueLoaderOption);
}
function setLoader$1(chain) {
    function customizer(object = '', sources = '') {
        if ([object, sources].every(e => typeof e === 'string'))
            return object + sources;
    }
    chain.plugin('mainPlugin')
        .tap(args => {
        args[0].loaderMeta = lodash.mergeWith(getLoaderMeta(), args[0].loaderMeta, customizer);
        return args;
    });
}
function setTaroApiLoader(chain) {
    chain.merge({
        module: {
            rule: {
                'process-import-taro-h5': {
                    test: /taro-h5[\\/]dist[\\/]api[\\/]taro/,
                    loader: require.resolve('./api-loader')
                }
            }
        }
    });
}

const CUSTOM_WRAPPER = 'custom-wrapper';
function modifyMiniWebpackChain(ctx, chain, data, config) {
    setVueLoader(ctx, chain, data, config);
    setLoader(chain);
    setDefinePlugin$1(chain);
}
function setVueLoader(ctx, chain, data, config) {
    var _a, _b, _c;
    var _d;
    const vueLoaderPath = getVueLoaderPath();
    // plugin
    const { VueLoaderPlugin } = require(vueLoaderPath);
    chain
        .plugin('vueLoaderPlugin')
        .use(VueLoaderPlugin);
    const compilerOptions = ((_a = config.vueLoaderOption) === null || _a === void 0 ? void 0 : _a.compilerOptions) || ((_b = config.mini) === null || _b === void 0 ? void 0 : _b.compilerOptions) || {};
    // loader
    const vueLoaderOption = Object.assign(Object.assign({ optimizeSSR: false, transformAssetUrls: {
            video: ['src', 'poster'],
            'live-player': 'src',
            audio: 'src',
            source: 'src',
            image: 'src',
            'cover-image': 'src'
        } }, ((_c = config.vueLoaderOption) !== null && _c !== void 0 ? _c : {})), { compilerOptions });
    (_d = vueLoaderOption.compilerOptions).nodeTransforms || (_d.nodeTransforms = []);
    vueLoaderOption.compilerOptions.nodeTransforms.unshift((node) => {
        if (node.type === 1 /* ELEMENT */) {
            node = node;
            const nodeName = node.tag;
            nodeName && ctx.applyPlugins({
                name: 'onParseCreateElement',
                opts: {
                    nodeName,
                    componentConfig: data.componentConfig
                }
            });
            if (template.capitalize(template.toCamelCase(nodeName)) in template.internalComponents) {
                // change only ElementTypes.COMPONENT to ElementTypes.ELEMENT
                // and leave ElementTypes.SLOT untouched
                if (node.tagType === 1 /* COMPONENT */) {
                    node.tagType = 0; /* ELEMENT */
                }
                // v-html
                const props = node.props;
                if (props.find(prop => prop.type === 7 && prop.name === 'html')) {
                    ['input', 'textarea', 'video', 'audio'].forEach(item => data.componentConfig.includes.add(item));
                }
                data.componentConfig.includes.add(nodeName);
            }
            if (nodeName === CUSTOM_WRAPPER) {
                node.tagType = 0; /* ELEMENT */
                data.componentConfig.thirdPartyComponents.set(CUSTOM_WRAPPER, new Set());
            }
            const usingComponent = data.componentConfig.thirdPartyComponents.get(nodeName);
            if (usingComponent != null) {
                node.props.forEach(prop => {
                    var _a;
                    if (prop.type === 6 /* ATTRIBUTE */) {
                        usingComponent.add(prop.name);
                    }
                    else if (prop.type === 7 /* DIRECTIVE */) {
                        prop = prop;
                        if (((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SimpleExpression */) {
                            let value = prop.arg.content;
                            if (prop.name === 'on') {
                                value = `on${value}`;
                            }
                            usingComponent.add(value);
                        }
                    }
                });
            }
        }
    });
    chain.module
        .rule('vue')
        .test(helper.REG_VUE)
        .use('vueLoader')
        .loader(vueLoaderPath)
        .options(vueLoaderOption);
}
function setLoader(chain) {
    chain.plugin('miniPlugin')
        .tap(args => {
        args[0].loaderMeta = getLoaderMeta();
        return args;
    });
}
function setDefinePlugin$1(chain) {
    chain
        .plugin('definePlugin')
        .tap(args => {
        args[0].ENABLE_ADJACENT_HTML = true;
        args[0].ENABLE_CLONE_NODE = true;
        args[0].ENABLE_TEMPLATE_CONTENT = true;
        args[0].ENABLE_MUTATION_OBSERVER = true;
        return args;
    });
}

var index = (ctx, config = {}) => {
    const { framework } = ctx.initialConfig;
    if (framework !== 'vue3')
        return;
    ctx.modifyWebpackChain(({ chain, data }) => {
        // 通用
        if (process.env.NODE_ENV !== 'production') {
            setAlias(chain);
        }
        setDefinePlugin(chain);
        if (process.env.TARO_PLATFORM === 'web') {
            // H5
            modifyH5WebpackChain(ctx, chain, config);
        }
        else {
            // 小程序
            modifyMiniWebpackChain(ctx, chain, data, config);
        }
    });
    ctx.modifyRunnerOpts(({ opts }) => {
        opts.frameworkExts = helper.VUE_EXT;
        if (!(opts === null || opts === void 0 ? void 0 : opts.compiler))
            return;
        if (shared.isString(opts.compiler)) {
            opts.compiler = {
                type: opts.compiler
            };
        }
        const { compiler } = opts;
        if (compiler.type === 'webpack5') {
            // 提供给 webpack5 依赖预编译收集器的第三方依赖
            const deps = [
                'vue',
                '@tarojs/plugin-framework-vue3/dist/runtime'
            ];
            compiler.prebundle || (compiler.prebundle = {});
            const prebundleOptions = compiler.prebundle;
            prebundleOptions.include || (prebundleOptions.include = []);
            prebundleOptions.include = prebundleOptions.include.concat(deps);
            const taroVue3Plugin = {
                name: 'taroVue3Plugin',
                setup(build) {
                    build.onLoad({ filter: /taro-h5[\\/]dist[\\/]api[\\/]taro/ }, ({ path }) => {
                        const content = helper.fs.readFileSync(path).toString();
                        return {
                            contents: require('./api-loader')(content)
                        };
                    });
                }
            };
            prebundleOptions.esbuild || (prebundleOptions.esbuild = {});
            const esbuildConfig = prebundleOptions.esbuild;
            esbuildConfig.plugins || (esbuildConfig.plugins = []);
            esbuildConfig.plugins.push(taroVue3Plugin);
        }
    });
    ctx.onParseCreateElement(({ nodeName, componentConfig }) => {
        if (template.capitalize(template.toCamelCase(nodeName)) in template.internalComponents) {
            componentConfig.includes.add(nodeName);
        }
    });
};
function setAlias(chain) {
    // 避免 npm link 时，taro composition apis 使用的 vue 和项目使用的 vue 实例不一致。
    chain.resolve.alias
        .set('vue', require.resolve('vue'));
}
function setDefinePlugin(chain) {
    chain
        .plugin('definePlugin')
        .tap(args => {
        const config = args[0];
        config.__VUE_OPTIONS_API__ = JSON.stringify(true);
        config.__VUE_PROD_DEVTOOLS__ = JSON.stringify(false);
        return args;
    });
}

exports.default = index;
