// @tarojs/plugin-framework-vue3/dist/runtime.js
import { isUndefined, isFunction, isArray, hooks, ensure } from '@tarojs/shared';
import { inject, ref, onMounted, onUnmounted, provide, defineComponent, shallowReactive, createApp, toRaw } from 'vue';
import { Current, getPageInstance, injectPageInstance, getPath, incrementId, safeExecute, eventHandler, addLeadingSlash, removePageInstance } from '@tarojs/runtime';

function createTaroHook(lifecycle) {
    return (fn) => {
        const id = inject('id');
        const fnRef = ref(fn);
        let inst;
        let callback;
        onMounted(() => {
            inst = getPageInstance(id);
            if (inst === undefined) {
                inst = Object.create({
                    $options: {}
                });
                injectPageInstance(inst, id);
            }
            inst = inst.$options;
            callback = (...args) => fnRef.value(...args);
            const currentCallback = inst[lifecycle];
            if (isUndefined(currentCallback)) {
                inst[lifecycle] = callback;
            }
            else if (isFunction(currentCallback)) {
                inst[lifecycle] = [inst[lifecycle], callback];
            }
            else if (isArray(currentCallback)) {
                inst[lifecycle] = [...currentCallback, callback];
            }
        });
        onUnmounted(() => {
            if (!inst || !callback) {
                return;
            }
            const list = inst[lifecycle];
            if (list === callback) {
                inst[lifecycle] = undefined;
            }
            else if (isArray(list)) {
                inst[lifecycle] = list.filter(item => item !== callback);
            }
            inst = null;
            callback = null;
        });
    };
}
/** LifeCycle */
const useDidShow = createTaroHook('onShow');
const useDidHide = createTaroHook('onHide');
/** App */
const useError = createTaroHook('onError');
const useUnhandledRejection = createTaroHook('onUnhandledRejection');
const useLaunch = createTaroHook('onLaunch');
const usePageNotFound = createTaroHook('onPageNotFound');
/** Page */
const useLoad = createTaroHook('onLoad');
const usePageScroll = createTaroHook('onPageScroll');
const usePullDownRefresh = createTaroHook('onPullDownRefresh');
const usePullIntercept = createTaroHook('onPullIntercept');
const useReachBottom = createTaroHook('onReachBottom');
const useResize = createTaroHook('onResize');
const useUnload = createTaroHook('onUnload');
/** Mini-Program */
const useAddToFavorites = createTaroHook('onAddToFavorites');
const useOptionMenuClick = createTaroHook('onOptionMenuClick');
const useSaveExitState = createTaroHook('onSaveExitState');
const useShareAppMessage = createTaroHook('onShareAppMessage');
const useShareTimeline = createTaroHook('onShareTimeline');
const useTitleClick = createTaroHook('onTitleClick');
/** Router */
const useReady = createTaroHook('onReady');
const useRouter = () => {
    // return dynamic ? Current.router : React.useMemo(() => Current.router, [])
    return Current.router;
};
const useTabItemTap = createTaroHook('onTabItemTap');

var taroHooks = /*#__PURE__*/Object.freeze({
  __proto__: null,
  useAddToFavorites: useAddToFavorites,
  useDidHide: useDidHide,
  useDidShow: useDidShow,
  useError: useError,
  useLaunch: useLaunch,
  useLoad: useLoad,
  useOptionMenuClick: useOptionMenuClick,
  usePageNotFound: usePageNotFound,
  usePageScroll: usePageScroll,
  usePullDownRefresh: usePullDownRefresh,
  usePullIntercept: usePullIntercept,
  useReachBottom: useReachBottom,
  useReady: useReady,
  useResize: useResize,
  useRouter: useRouter,
  useSaveExitState: useSaveExitState,
  useShareAppMessage: useShareAppMessage,
  useShareTimeline: useShareTimeline,
  useTabItemTap: useTabItemTap,
  useTitleClick: useTitleClick,
  useUnhandledRejection: useUnhandledRejection,
  useUnload: useUnload
});

const setGlobalDataPlugin = {
    install: (app, data) => {
        app.taroGlobalData = data;
    }
};

/**
 * set writable, enumerable to true
 */
function setDefaultDescriptor(obj) {
    obj.writable = true;
    obj.enumerable = true;
    return obj;
}
/**
 * 设置入口的路由参数
 * @param options 小程序传入的参数
 */
function setRouterParams(options) {
    Current.router = Object.assign({ params: options === null || options === void 0 ? void 0 : options.query }, options);
}

function setReconciler() {
    hooks.tap('getLifecycle', function (instance, lifecycle) {
        return instance.$options[lifecycle];
    });
    if (process.env.TARO_PLATFORM === 'web') {
        hooks.tap('createPullDownComponent', (component, path, h, _, stampId) => {
            const inject = {
                props: {
                    tid: String
                },
                created() {
                    const pagePath = stampId ? getPath(path, { stamp: stampId }) : path;
                    injectPageInstance(this, pagePath);
                }
            };
            component.mixins = isArray(component.mixins)
                ? component.mixins.push(inject)
                : [inject];
            return {
                render() {
                    return h('taro-pull-to-refresh-core', {
                        class: 'hydrated'
                    }, [h(component, this.$slots.default)]);
                }
            };
        });
        hooks.tap('getDOMNode', el => {
            return el.$el;
        });
    }
}
function createVue3Page(h, id) {
    return function (component) {
        var _a, _b, _c, _d;
        // 处理类组件
        component = isClassComponent(component) ? component.__vccOpts : component;
        const inject = {
            props: {
                tid: String
            },
            created() {
                injectPageInstance(this, id);
            }
        };
        if (isArray(component.mixins)) {
            const mixins = component.mixins;
            const idx = mixins.length - 1;
            if (!((_a = mixins[idx].props) === null || _a === void 0 ? void 0 : _a.tid)) {
                // mixins 里还没注入过，直接推入数组
                component.mixins.push(inject);
            }
            else {
                // mixins 里已经注入过，代替前者
                component.mixins[idx] = inject;
            }
        }
        else {
            component.mixins = [inject];
        }
        const ProviderComponent = {
            setup() {
                provide('id', id);
            },
            render() {
                return this.$slots.default();
            }
        };
        const RootElement = process.env.TARO_PLATFORM === 'web' ? 'div' : 'root';
        const PageComponent = Object.assign({}, component);
        const option = ((_d = (_c = (_b = PageComponent.props) === null || _b === void 0 ? void 0 : _b.option) === null || _c === void 0 ? void 0 : _c.default) === null || _d === void 0 ? void 0 : _d.call(_c)) || {};
        return h(ProviderComponent, {
            key: id
        }, {
            default() {
                return [
                    h(RootElement, {
                        id,
                        class: process.env.TARO_PLATFORM === 'web' ? 'taro_page' : ''
                    }, [
                        h(PageComponent, { tid: id, option })
                    ])
                ];
            }
        });
    };
}
function createVue3App(app, h, config) {
    let pages = [];
    let appInstance;
    ensure(!(isFunction(app._component) && !isClassComponent(app._component)), '入口组件不支持使用函数式组件');
    setReconciler();
    app._component.render = function () {
        return pages.slice();
    };
    if (process.env.TARO_PLATFORM !== 'web') {
        appInstance = app.mount('#app');
    }
    const [ONLAUNCH, ONSHOW, ONHIDE] = hooks.call('getMiniLifecycleImpl').app;
    const appConfig = Object.create({
        mount(component, id, cb) {
            const page = createVue3Page(h, id)(component);
            pages.push(page);
            this.updateAppInstance(cb);
        },
        unmount(id, cb) {
            pages = pages.filter(page => page.key !== id);
            this.updateAppInstance(cb);
        },
        updateAppInstance(cb) {
            appInstance.$forceUpdate();
            appInstance.$nextTick(cb);
        }
    }, {
        config: setDefaultDescriptor({
            configurable: true,
            value: config
        }),
        [ONLAUNCH]: setDefaultDescriptor({
            value(options) {
                var _a;
                setRouterParams(options);
                if (process.env.TARO_PLATFORM === 'web') {
                    appInstance = app.mount(`#${config.appId || 'app'}`);
                }
                // 把 App Class 上挂载的额外属性同步到全局 app 对象中
                // eslint-disable-next-line dot-notation
                if (app['taroGlobalData']) {
                    // eslint-disable-next-line dot-notation
                    const globalData = app['taroGlobalData'];
                    const keys = Object.keys(globalData);
                    const descriptors = Object.getOwnPropertyDescriptors(globalData);
                    keys.forEach(key => {
                        Object.defineProperty(this, key, {
                            configurable: true,
                            enumerable: true,
                            get() {
                                return globalData[key];
                            },
                            set(value) {
                                globalData[key] = value;
                            }
                        });
                    });
                    Object.defineProperties(this, descriptors);
                }
                const onLaunch = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onLaunch;
                isFunction(onLaunch) && onLaunch.call(appInstance, options);
            }
        }),
        [ONSHOW]: setDefaultDescriptor({
            value(options) {
                var _a;
                setRouterParams(options);
                const onShow = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onShow;
                isFunction(onShow) && onShow.call(appInstance, options);
            }
        }),
        [ONHIDE]: setDefaultDescriptor({
            value(options) {
                var _a;
                const onHide = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onHide;
                isFunction(onHide) && onHide.call(appInstance, options);
            }
        }),
        onError: setDefaultDescriptor({
            value(error) {
                var _a;
                const onError = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onError;
                isFunction(onError) && onError.call(appInstance, error);
            }
        }),
        onUnhandledRejection: setDefaultDescriptor({
            value(error) {
                var _a;
                const onUnhandledRejection = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onUnhandledRejection;
                isFunction(onUnhandledRejection) && onUnhandledRejection.call(appInstance, error);
            }
        }),
        onPageNotFound: setDefaultDescriptor({
            value(res) {
                var _a;
                const onPageNotFound = (_a = appInstance === null || appInstance === void 0 ? void 0 : appInstance.$options) === null || _a === void 0 ? void 0 : _a.onPageNotFound;
                isFunction(onPageNotFound) && onPageNotFound.call(appInstance, res);
            }
        })
    });
    Current.app = appConfig;
    return appConfig;
}
function isClassComponent(value) {
    return isFunction(value) && '__vccOpts' in value;
}

const getNativeCompId = incrementId();
function initNativeComponentEntry(h) {
    setReconciler();
    const NativeComponentWrapper = {
        props: ['getCtx', 'compId'],
        setup(props) {
            const root = ref();
            const ctx = props.getCtx();
            provide('id', props.compId);
            onMounted(() => {
                const rootElement = toRaw(root.value);
                rootElement.ctx = ctx;
                rootElement.performUpdate(true);
            });
            return { root };
        },
        render() {
            return h('root', {
                ref: 'root',
                id: this.compId
            }, this.$slots.default());
        }
    };
    const App = defineComponent({
        setup() {
            const components = shallowReactive([]);
            function mount(component, compId, getCtx) {
                var _a, _b, _c, _d;
                // 处理类组件
                component = isClassComponent(component) ? component.__vccOpts : component;
                const inject = {
                    props: {
                        tid: String
                    },
                    created() {
                        injectPageInstance(this, compId);
                    }
                };
                if (isArray(component.mixins)) {
                    const mixins = component.mixins;
                    const idx = mixins.length - 1;
                    if (!((_a = mixins[idx].props) === null || _a === void 0 ? void 0 : _a.tid)) {
                        // mixins 里还没注入过，直接推入数组
                        component.mixins.push(inject);
                    }
                    else {
                        // mixins 里已经注入过，代替前者
                        component.mixins[idx] = inject;
                    }
                }
                else {
                    component.mixins = [inject];
                }
                const PageComponent = Object.assign({}, component);
                const option = ((_d = (_c = (_b = PageComponent.props) === null || _b === void 0 ? void 0 : _b.option) === null || _c === void 0 ? void 0 : _c.default) === null || _d === void 0 ? void 0 : _d.call(_c)) || {};
                const ctx = getCtx();
                const page = {
                    mounted() {
                        ctx.component = this;
                    },
                    unmounted() {
                        removePageInstance(compId);
                    },
                    render() {
                        return h(NativeComponentWrapper, {
                            compId,
                            getCtx
                        }, {
                            default() {
                                return [
                                    h(PageComponent, Object.assign(Object.assign({ tid: compId, option }, (ctx.data || (ctx.data = {})).props), { _scope: ctx }))
                                ];
                            }
                        });
                    }
                };
                components.push({
                    compId,
                    component: page
                });
            }
            function unmount(compId) {
                const index = components.findIndex(page => page.compId === compId);
                components.splice(index, 1);
            }
            onMounted(() => (Current.app = {
                mount: mount,
                unmount
            }));
            return {
                components
            };
        },
        render() {
            return this.components.map(page => h(page.component, { key: page.compId }));
        }
    });
    createApp(App).mount('#app');
}
function createNativeComponentConfig(component, h, componentConfig) {
    const componentObj = {
        options: componentConfig,
        properties: {
            props: {
                type: null,
                value: null,
                observer(_newVal, oldVal) {
                    var _a;
                    oldVal && ((_a = this.component) === null || _a === void 0 ? void 0 : _a.$forceUpdate());
                }
            }
        },
        created() {
            if (!Current.app) {
                initNativeComponentEntry(h);
            }
        },
        attached() {
            const compId = this.compId = getNativeCompId();
            setCurrent(compId);
            this.config = componentConfig;
            Current.app.mount(component, compId, () => this);
        },
        ready() {
            safeExecute(this.compId, 'onReady');
        },
        detached() {
            resetCurrent();
            Current.app.unmount(this.compId);
            this.component = null;
        },
        pageLifetimes: {
            show(options) {
                safeExecute(this.compId, 'onShow', options);
            },
            hide() {
                safeExecute(this.compId, 'onHide');
            }
        },
        methods: {
            eh: eventHandler
        }
    };
    function resetCurrent() {
        // 小程序插件页面卸载之后返回到宿主页面时，需重置Current页面和路由。否则引发插件组件二次加载异常 fix:#11991
        Current.page = null;
        Current.router = null;
    }
    function setCurrent(compId) {
        const pages = getCurrentPages();
        const currentPage = pages[pages.length - 1];
        if (Current.page === currentPage)
            return;
        Current.page = currentPage;
        const route = currentPage.route || currentPage.__route__;
        const router = {
            params: currentPage.options || {},
            path: addLeadingSlash(route),
            $taroPath: compId,
            onReady: '',
            onHide: '',
            onShow: ''
        };
        Current.router = router;
        if (!currentPage.options) {
            // 例如在微信小程序中，页面 options 的设置时机比组件 attached 慢
            Object.defineProperty(currentPage, 'options', {
                enumerable: true,
                configurable: true,
                get() {
                    return this._optionsValue;
                },
                set(value) {
                    router.params = value;
                    this._optionsValue = value;
                }
            });
        }
    }
    return componentObj;
}

hooks.tap('initNativeApi', function (taro) {
    for (const hook in taroHooks) {
        taro[hook] = taroHooks[hook];
    }
    taro.setGlobalDataPlugin = setGlobalDataPlugin;
});
hooks.tap('proxyToRaw', function (proxyObj) {
    return toRaw(proxyObj);
});

export { createNativeComponentConfig, createVue3App, isClassComponent, setGlobalDataPlugin, setReconciler, useAddToFavorites, useDidHide, useDidShow, useError, useLaunch, useLoad, useOptionMenuClick, usePageNotFound, usePageScroll, usePullDownRefresh, usePullIntercept, useReachBottom, useReady, useResize, useRouter, useSaveExitState, useShareAppMessage, useShareTimeline, useTabItemTap, useTitleClick, useUnhandledRejection, useUnload };
